### PR TITLE
Bug fix for update_weights_from_tensor when pp_size > 1

### PR DESCRIPTION
**Pull request by WanXiaopei**
## Modifications

We are the team in AntGroup. When we employ sglang as the rollout backend in verl, we found than if we set the pp_size > 1, then OOM was occured after we called update_weghts_from_tensor. 

Then we found that sglang has not supported update_weghts_from_tensor under pp_size > 1, and the issue https://github.com/sgl-project/sglang/issues/8076 we proposed before is related to this problem.

Here is our example for calling update_weights_from_tensor under pp_size > 1.

```python
# Copyright (c) 2025 RedAccel Authors. All Rights Reserved.

import argparse
import os
import tqdm
import asyncio
from typing import Iterable, Tuple, List, Optional
from itertools import islice

from loguru import logger
from sglang.srt.model_executor.model_runner import LocalSerializedTensor
from sglang.srt.utils import MultiprocessingSerializer
import torch
import torch.distributed as dist
from transformers import AutoModelForCausalLM
from sglang.srt.entrypoints.engine import Engine
from sglang.srt.managers.io_struct import UpdateWeightsFromTensorReqInput

torch.multiprocessing.set_start_method('spawn', force=True)
parser = argparse.ArgumentParser()
parser.add_argument("--model", type=str)
parser.add_argument("--chunk_size", type=int, default=1)
args = parser.parse_args()


class AsyncEngine(Engine):
    def update_weights_from_tensor(
        self,
        named_tensors: List[Tuple[str, torch.Tensor]],
        load_format: Optional[str] = None,
        flush_cache: bool = True,
    ):
        """Update weights from distributed source. If there are going to be more updates, set `flush_cache` to be false
        to avoid duplicated cache cleaning operation."""
        obj = UpdateWeightsFromTensorReqInput(
            serialized_named_tensors=named_tensors,
            load_format=load_format,
            flush_cache=flush_cache,
        )
        loop = asyncio.get_event_loop()
        return loop.run_until_complete(
            self.tokenizer_manager.update_weights_from_tensor(obj, None)
        )


def batched(iterable, n):
    # batched('ABCDEFG', 3) --> ABC DEF G
    if n < 1:
        raise ValueError("n must be at least one")
    it = iter(iterable)
    while batch := tuple(islice(it, n)):
        yield batch


def per_tensor_generator(model) -> Iterable[Tuple[str, torch.Tensor]]:
    for name, tensor in model.named_parameters():
        yield name, tensor.to(torch.cuda.current_device())


def main():
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    world_size = dist.get_world_size()
    torch.cuda.set_device(rank)

    os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
    os.environ["NCCL_CUMEM_ENABLE"] = "0"
    os.environ["TORCH_NCCL_AVOID_RECORD_STREAMS"] = "1"
    os.environ["CUDA_DEVICE_MAX_CONNECTIONS"] = "4"
    os.environ["CUDA_MODULE_LOADING"] = "AUTO"
    for k in ["TORCHELASTIC_USE_AGENT_STORE"]:
        if k in os.environ:
            del os.environ[k]
    from torch.distributed.device_mesh import init_device_mesh
    tp_size = world_size // 2
    pp_size = 2
    device_mesh = init_device_mesh("cuda", mesh_shape=(pp_size, tp_size), mesh_dim_names=("pp", "tp"))

    pp_rank = device_mesh["pp"].get_local_rank()
    tp_rank = device_mesh["tp"].get_local_rank()
    print(f"{dist.get_rank()=}, {pp_rank=}, {tp_rank=}")

    if device_mesh["pp"].get_local_rank() == 0 and device_mesh["tp"].get_local_rank() == 0:
        os.environ["SGLANG_BLOCK_NONZERO_RANK_CHILDREN"] = "0"
        inference_engine = AsyncEngine(
            model_path=args.model,
            dtype="bfloat16",
            mem_fraction_static=0.4,
            enable_memory_saver=True,
            tp_size=world_size // 2,
            pp_size=2,
            node_rank=0,
            nnodes=1,
            load_format="dummy",
            dist_init_addr=None,
            trust_remote_code=True,
            mm_attention_backend="fa3",
            attention_backend="fa3",
            disable_cuda_graph=True,
            grammar_backend="outlines",
        )

    dist.barrier()

    model = AutoModelForCausalLM.from_pretrained(
        args.model,
        device_map="cpu",
        trust_remote_code=True,
        low_cpu_mem_usage=False,
    )
    tensor_num = len(model.state_dict())

    def update_weights_by_chunk(named_tensors):
        logger.info(
            f"Rank {rank} start to update weights with {tensor_num=} and {args.chunk_size=}"
        )
        pbar = tqdm.tqdm(total=tensor_num)
        for chunk in batched(named_tensors, args.chunk_size):
            serialized_tensor = [
                (name, MultiprocessingSerializer.serialize(tensor.detach()))
                for name, tensor in chunk
            ]

            tp_gathered_serialized_tensors = [None for _ in range(tp_size)]

            dist.all_gather_object(
                obj=serialized_tensor,
                object_list=tp_gathered_serialized_tensors,
                group=device_mesh.get_group("tp")
            )

            tp_gathered_serialized_tensors = [
                (i[0][0], LocalSerializedTensor(values=[j[1] for j in i]))
                for i in zip(*tp_gathered_serialized_tensors)
            ]

            if device_mesh["pp"].get_local_rank() == 0:
                pp_gathered_serialized_tensors = [None for _ in range(pp_size)]
            else:
                pp_gathered_serialized_tensors = None

            dist.gather_object(
                obj=tp_gathered_serialized_tensors,
                object_gather_list=pp_gathered_serialized_tensors,
                dst=device_mesh["pp"].mesh.tolist()[0],
                group=device_mesh.get_group("pp")
            )

            if device_mesh["pp"].get_local_rank() == 0 and device_mesh["tp"].get_local_rank() == 0:
                pp_gathered_serialized_tensors = [
                    [MultiprocessingSerializer.serialize(tp_tensors) for _ in range(tp_size)]
                    for tp_tensors in pp_gathered_serialized_tensors
                ]

                inference_engine.update_weights_from_tensor(
                    named_tensors=pp_gathered_serialized_tensors,
                    load_format=None,
                    flush_cache=False,
                )
            pbar.update(len(chunk))
        dist.barrier()

    named_tensors = per_tensor_generator(model)
    update_weights_by_chunk(named_tensors)
    if rank == 0:
        tokenizer = inference_engine.tokenizer_manager.tokenizer
        prompt = [{"role": "system", "content": "You are a helpful assistant."},
                  {"role": "user", "content": "介绍一下四川省"}]
        prompt = tokenizer.apply_chat_template(prompt, tokenize=False, add_generation_prompt=True, )
        ret = inference_engine.generate(prompt=prompt)
        print(ret)
        inference_engine.shutdown()


if __name__ == "__main__":
    main()

```

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/managers/scheduler.py [hunk 1]
```diff
@@ -2499,7 +2499,33 @@ def update_weights_from_distributed(
...
+             List[List[Union[LocalSerializedTensor, Tensor]]]. The first dimension indicates the pipeline parallelism,
+             and the second dimension indicates the tensor parallelism. Therefore, when this process receives a
+             request of UpdateWeightsFromTensorReqInput, then it must select the serialized_named_tensors according to
+             its pp_rank. Only in this way, we can avoid CUDA OOM issue caused by deserialize a tensor in rank `m` which is
+             ForkingPickler serialized in rank `n`. where `m` is not equal to `n`.
+        """
+        if isinstance(recv_req.serialized_named_tensors[0], List):
+            if self.pp_size != len(recv_req.serialized_named_tensors):
+                message = (f"The size of serialized_named_tensors in UpdateWeightsFromTensorReqInput must be equal to "
+                           f"the size of pp_size, but got {len(recv_req.serialized_named_tensors)} vs. {self.pp_size}")
```
[the remark this answers is not in the document]
Comment (WanXiaopei) [reply python/sglang/srt/managers/scheduler.py:2432]: misreport bug

File: python/sglang/srt/managers/scheduler.py [hunk 2]
```diff
@@ -2499,7 +2499,33 @@ def update_weights_from_distributed(
...
+             ForkingPickler, and represents the tensor from each tensor parallel rank.
+
+           When pp_size is greater than 1, then recv_req.serialized_named_tensors must a data structure with type of
+             List[List[Union[LocalSerializedTensor, Tensor]]]. The first dimension indicates the pipeline parallelism,
+             and the second dimension indicates the tensor parallelism. Therefore, when this process receives a
+             request of UpdateWeightsFromTensorReqInput, then it must select the serialized_named_tensors according to
+             its pp_rank. Only in this way, we can avoid CUDA OOM issue caused by deserialize a tensor in rank `m` which is
+             ForkingPickler serialized in rank `n`. where `m` is not equal to `n`.
+        """
+        if isinstance(recv_req.serialized_named_tensors[0], List):
```
[the remark this answers is not in the document]
Comment (WanXiaopei) [reply python/sglang/srt/managers/scheduler.py]: fixed
